### PR TITLE
Node should start up despite of a lingering `.es_temp_file`

### PR DESCRIPTION
**Pull request by bleskes**
When ES starts up we verify we can write to all data files by creating and deleting a temp file called `.es_temp_file`. If for some reason the file was successfully created but not successfully deleted, we still shut down correctly but subsequent start attempts will fail with a file already exists exception.

This PR makes sure to first clean any existing `.es_temp_file`

Superseeds #21007

### CODE REVIEW

**Comment by bleskes** [comment]
@s1monw can you take a look please?

**Comment by jasontedor** [comment]
If we are going to make this change, I think we should make the same change for the atomic move temp file.

**Comment by bleskes** [comment]
@jasontedor agreed. Pushed another commit

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -891,11 +894,12 @@ public void ensureAtomicMoveSupported() throws IOException {
         final NodePath[] nodePaths = nodePaths();
         for (NodePath nodePath : nodePaths) {
             assert Files.isDirectory(nodePath.path) : nodePath.path + " is not a directory";
-            final Path src = nodePath.path.resolve("__es__.tmp");
-            final Path target = nodePath.path.resolve("__es__.final");
+            final Path src = nodePath.path.resolve(TEMP_FILE_NAME + ".src");
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: maybe don't change the name if you want to delete it first

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 2]
```diff
@@ -999,19 +1003,19 @@ private void assertCanWrite() throws IOException {
...
+    static final String TEMP_FILE_NAME = ".es_temp_file";
+
     private static void tryWriteTempFile(Path path) throws IOException {
         if (Files.exists(path)) {
-            Path resolve = path.resolve(".es_temp_file");
-            boolean tempFileCreated = false;
+            Path resolve = path.resolve(TEMP_FILE_NAME);
             try {
+                // delete any lingering file from a previous failure
+                Files.deleteIfExists(resolve);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: maybe move all of the non-write operations out of the try block then the exception makes sense again

File: core/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 3]
```diff
@@ -891,11 +894,12 @@ public void ensureAtomicMoveSupported() throws IOException {
         final NodePath[] nodePaths = nodePaths();
         for (NodePath nodePath : nodePaths) {
             assert Files.isDirectory(nodePath.path) : nodePath.path + " is not a directory";
-            final Path src = nodePath.path.resolve("__es__.tmp");
-            final Path target = nodePath.path.resolve("__es__.final");
+            final Path src = nodePath.path.resolve(TEMP_FILE_NAME + ".src");
+            final Path target = nodePath.path.resolve(TEMP_FILE_NAME + ".target");
             try {
+                Files.deleteIfExists(src);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: if we do this, then we can also just do `if (Files.exists(src))` here?

**Comment by s1monw** [comment]
left some comments...  I really wonder if we should do this by default or if we should do this only if somebody specifies a cmd option / system property?

**Comment by bleskes** [comment]
> I really wonder if we should do this by default or if we should do this only if somebody specifies a cmd option / system property?

I think it's good for these tests not to have side effects on failures? what's your concern?

**Comment by s1monw** [comment]
> I think it's good for these tests not to have side effects on failures? what's your concern?

the failure is what a complete node going down due to powerloss? Pretty rare... I am concerned we hide bugs if we just go and add such a leniency

**Comment by bleskes** [comment]
> I am concerned we hide bugs if we just go and add such a leniency

Not sure I follow - we want to make sure that we can write to those folders? I think potentially deleting files in those folders is inline with the check and just strengthen it?

**Comment by s1monw** [comment]
> Not sure I follow - we want to make sure that we can write to those folders? I think potentially deleting files in those folders is inline with the check and just strengthen it?

we have a precondition that is not met, if it's there it's an exceptional situation and potentially a bug.

**Comment by perlei** [comment]
Im experiencing this with elasticsearch 5.1.1 sometimes when i restart the VM that elasticsearch is running on, it can not start i get the following : 
```
 Caused by: java.nio.file.FileAlreadyExistsException: /mnt/data/elasticsearch/data/nodes/0/indices/dZ4wmoCmQtCM-PiD7pn8QQ/2/_state/.es_temp_file
 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:88) ~[?:?]
 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102) ~[?:?]
 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107) ~[?:?]
 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214) ~[?:?]
 	at java.nio.file.Files.newByteChannel(Files.java:361) ~[?:1.8.0_112]
 	at java.nio.file.Files.createFile(Files.java:632) ~[?:1.8.0_112]
 	at org.elasticsearch.env.NodeEnvironment.tryWriteTempFile(NodeEnvironment.java:1033) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.env.NodeEnvironment.assertCanWrite(NodeEnvironment.java:1019) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.env.NodeEnvironment.<init>(NodeEnvironment.java:276) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.node.Node.<init>(Node.java:249) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.node.Node.<init>(Node.java:229) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.bootstrap.Bootstrap$6.<init>(Bootstrap.java:214) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.bootstrap.Bootstrap.setup(Bootstrap.java:214) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.bootstrap.Bootstrap.init(Bootstrap.java:306) ~[elasticsearch-5.1.1.jar:5.1.1]
 	at org.elasticsearch.bootstrap.Elasticsearch.init(Elasticsearch.java:121) ~[elasticsearch-5.1.1.jar:5.1.1]

```

My workaround is to add the following line: 
´´´
find /mnt/data/elasticsearch/data/ -name .es_temp_file -print | xargs -n 1 rm -rf
´´´

To the elasticsearch startup script.

**Comment by s1monw** [comment]
@bleskes I think I changed my mind, lets do this. The reason why I think we should do it is that if you press ctrl+c while stuff is starting up we might leak the file. It's ok to delete it first since we retry to write it immediately and that is it's only purpose, to ensure we can write.

**Review by s1monw — APPROVED** [review]

**Pull request merged**